src/main/java/bio/terra/workspace/service/resource/controlled/flight/create/CreateBigQueryDatasetStep.java

```diff
@@ -0,0 +1,183 @@
+
+    return StepResult.getStepResultSuccess();
+  }
+
+  /**
+   * Builds the IAM configuration to grant workspace users access to the new dataset.
+   *
+   * <p>Unlike other cloud objects, BQ datasets do not use GCP's common IAM objects. Instead, they
+   * use Access objects, which are roughly equivalent to the Bindings used elsewhere. With some
+   * translation, that means this can still use the resource-type-agnostic policy-building code from
+   * {@link GcpPolicyBuilder}.
+   */
+  private List<Access> buildDatasetIamConfiguration(FlightMap workingMap) {
+    // As this is a new dataset, we pass an empty Policy object as the initial state to
+    // GcpPolicyBuilder.
+    GcpPolicyBuilder policyBuilder =
+        new GcpPolicyBuilder(resource, resource.getProjectId(), Policy.newBuilder().build());
+
+    // Read Sam groups for each workspace role. Stairway does not
+    // have a cleaner way of deserializing parameterized types, so we suppress warnings here.
+    @SuppressWarnings("unchecked")
+    Map<WsmIamRole, String> workspaceRoleGroupsMap =
+        workingMap.get(WorkspaceFlightMapKeys.IAM_GROUP_EMAIL_MAP, Map.class);
+    for (Map.Entry<WsmIamRole, String> entry : workspaceRoleGroupsMap.entrySet()) {
+      policyBuilder.addWorkspaceBinding(entry.getKey(), entry.getValue());
+    }
+
+    // Resources with permissions given to individual users (private or application managed) use
+    // the resource's Sam policies to manage those individuals, so they must be synced here.
+    // This section should also run for application managed resources, once those are supported.
+    if (resource.getAccessScope() == AccessScopeType.ACCESS_SCOPE_PRIVATE) {
+      @SuppressWarnings("unchecked")
+      Map<ControlledResourceIamRole, String> resourceRoleGroupsMap =
+          workingMap.get(ControlledResourceKeys.IAM_RESOURCE_GROUP_EMAIL_MAP, Map.class);
+      for (Map.Entry<ControlledResourceIamRole, String> entry : resourceRoleGroupsMap.entrySet()) {
+        policyBuilder.addResourceBinding(entry.getKey(), entry.getValue());
+      }
+    }
+
+    Policy updatedPolicy = policyBuilder.build();
+    List<Binding> bindingList = updatedPolicy.getBindingsList();
+    return bindingList.stream().map(this::toAccess).collect(Collectors.toList());
+  }
+
+  /**
+   * Translate a Binding object to an Access object.
+   *
+   * <p>Binding objects are a common IAM construct used across many GCP libraries. They represent a
+   * single GCP role applied to one or more members.
+   *
+   * <p>Access objects are a BigQuery-specific IAM construct. They represent a single GCP role
+   * applied to a single member. For legacy reasons, BigQuery datasets use lists of Access objects
+   * instead of lists of Binding objects for IAM control.
+   *
+   * <p>This translation assumes that the provided Binding object only has a single member. This
+   * holds for all Bindings built by a GcpPolicyBuilder object.
+   */
+  private Access toAccess(Binding binding) {
```

Comment:
a bit yeah - for most other resources, create + cloud sync is 2 steps instead of 1, and this would normally live inside something like `BigQueryDatasetCloudSyncStep`. It seems reasonable to combine them here and there should be enough comments to explain why... but yeah, it's a big step.